core/trino-main/src/main/java/io/trino/execution/scheduler/PipelinedStageExecution.java

```diff
@@ -390,8 +390,7 @@ private void updateTaskStatus(TaskStatus taskStatus)
     private synchronized boolean isStageFlushing()
     {
         // to transition to flushing, there must be at least one flushing task, and all others must be flushing or finished.
-        return !flushingTasks.isEmpty()
-                && allTasks.stream().allMatch(taskId -> finishedTasks.contains(taskId) || flushingTasks.contains(taskId));
+        return !flushingTasks.isEmpty() && allTasks.size() == finishedTasks.size() + flushingTasks.size();
```

Comment:
Should there be a checkState that `finishedTasks.containsAll(allTasks) && finishedTasks.containsAll(flushingTasks)` for consistency with isStageFinished ?

Reply:
`addFinishedTasks` first is adding task to finished set and then removes it from flushing task. So there is no possibility that one task will be counted in both sets (added to finished, not yet removed from flushing)? Then this will fail to transform stage to flushing.

Reply:
> So there is no possibility that one task will be counted in both sets (added to finished, not yet removed from flushing)?

Task can only be in one set. `flushingTasks` and `finishedTasks` are still under `@GuardedBy(this)` lock and locked opportunistally.

> Should there be a checkState that finishedTasks.containsAll(allTasks) && finishedTasks.containsAll(flushingTasks) for consistency with isStageFinished ?

You would have to run this check every time task transitions from `flushing` to `finished`. I don't think it's needed